File: trading-app-frontend/src/services/appwriteRealtime.js
/**
 * Appwrite Realtime Service
 * Handles real-time subscriptions, live updates, and notifications
 */

import { client, DATABASE_ID, COLLECTIONS, handleAppwriteError } from '../lib/appwrite';

class AppwriteRealtimeService {
  constructor() {
    this.subscriptions = new Map();
    this.eventHandlers = new Map();
    this.isConnected = false;
  }

  // Subscribe to real-time events
  subscribe(channels, callback, subscriptionId = null) {
    try {
      const id = subscriptionId || `sub_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
      
      console.log('🔔 Setting up real-time subscription:', { id, channels });
      
      // Create subscription
      const unsubscribe = client.subscribe(channels, (response) => {
        try {
          console.log('📡 Real-time event received:', response);
          
          const eventData = {
            event: response.event || response.events?.[0],
            payload: response.payload,
            timestamp: new Date().toISOString(),
            channels: response.channels,
            subscriptionId: id
          };
          
          // Call the callback with processed event data
          callback(eventData);
          
        } catch (error) {
          console.error('❌ Real-time callback error:', error);
        }
      });
      
      // Store subscription
      this.subscriptions.set(id, {
        channels,
        callback,
        unsubscribe,
        createdAt: new Date().toISOString()
      });
      
      console.log('✅ Real-time subscription created:', id);
      return id;
      
    } catch (error) {
      console.error('❌ Real-time subscription failed:', error);
      return null;
    }
  }

  // Unsubscribe from real-time events
  unsubscribe(subscriptionId) {
    try {
      const subscription = this.subscriptions.get(subscriptionId);
      
      if (subscription) {
        // Call the unsubscribe function
        subscription.unsubscribe();
        
        // Remove from subscriptions
        this.subscriptions.delete(subscriptionId);
        
        console.log('✅ Unsubscribed from real-time events:', subscriptionId);
        return true;
      }
      
      console.warn('⚠️ Subscription not found:', subscriptionId);
      return false;
      
    } catch (error) {
      console.error('❌ Unsubscribe failed:', error);
      return false;
    }
  }

  // Unsubscribe from all active subscriptions
  unsubscribeAll() {
    try {
      let count = 0;
      
      for (const [id, subscription] of this.subscriptions) {
        try {
          subscription.unsubscribe();
          count++;
        } catch (error) {
          console.error(`❌ Failed to unsubscribe ${id}:`, error);
        }
      }
      
      this.subscriptions.clear();
      this.eventHandlers.clear();
      
      console.log(`✅ Unsubscribed from ${count} real-time subscriptions`);
      return count;
      
    } catch (error) {
      console.error('❌ Unsubscribe all failed:', error);
      return 0;
    }
  }

  // Subscribe to user-specific messages
  subscribeToMessages(userId, callback) {
    const channels = [
      `databases.${DATABASE_ID}.collections.${COLLECTIONS.messages}.documents`,
    ];
    
    const messageCallback = (eventData) => {
      const { event, payload } = eventData;
      
      // Only process message events for this user
      if (event && event.includes('.messages.') && payload) {
        // Check if user is the recipient
        if (payload.recipient_id === userId) {
          callback({
            type: 'new_message',
            message: payload,
            ...eventData
          });
        }
      }
    };
    
    return this.subscribe(channels, messageCallback, `messages_${userId}`);
  }

  // Subscribe to user trades/matches
  subscribeToTrades(userId, callback) {
    const channels = [
      `databases.${DATABASE_ID}.collections.${COLLECTIONS.trades}.documents`,
    ];
    
    const tradeCallback = (eventData) => {
      const { event, payload } = eventData;
      
      // Only process trade events for this user
      if (event && event.includes('.trades.') && payload) {
        // Check if user is involved in this trade
        if (payload.user1_id === userId || payload.user2_id === userId) {
          let eventType = 'trade_updated';
          
          if (event.includes('.create')) {
            eventType = 'new_trade';
          } else if (event.includes('.update')) {
            eventType = 'trade_updated';
          } else if (event.includes('.delete')) {
            eventType = 'trade_deleted';
          }
          
          callback({
            type: eventType,
            trade: payload,
            ...eventData
          });
        }
      }
    };
    
    return this.subscribe(channels, tradeCallback, `trades_${userId}`);
  }

  // Subscribe to marketplace updates
  subscribeToMarketplace(callback, filters = {}) {
    const channels = [
      `databases.${DATABASE_ID}.collections.${COLLECTIONS.items}.documents`,
    ];
    
    const marketplaceCallback = (eventData) => {
      const { event, payload } = eventData;
      
      if (event && event.includes('.items.') && payload) {
        // Apply filters
        if (filters.location && payload.location !== filters.location) {
          return;
        }
        
        if (filters.category && payload.category !== filters.category) {
          return;
        }
        
        let eventType = 'listing_updated';
        
        if (event.includes('.create')) {
          eventType = 'new_listing';
        } else if (event.includes('.update')) {
          eventType = 'listing_updated';
        } else if (event.includes('.delete')) {
          eventType = 'listing_deleted';
        }
        
        callback({
          type: eventType,
          listing: payload,
          ...eventData
        });
      }
    };
    
    const filterId = `marketplace_${filters.location || 'all'}_${filters.category || 'all'}`;
    return this.subscribe(channels, marketplaceCallback, filterId);
  }

  // Subscribe to user profile updates
  subscribeToUserUpdates(userId, callback) {
    const channels = [
      `databases.${DATABASE_ID}.collections.${COLLECTIONS.users}.documents.${userId}`,
    ];
    
    const userCallback = (eventData) => {
      const { event, payload } = eventData;
      
      if (event && event.includes('.users.') && payload) {
        callback({
          type: 'profile_updated',
          profile: payload,
          ...eventData
        });
      }
    };
    
    return this.subscribe(channels, userCallback, `user_${userId}`);
  }

  // Subscribe to AI matching updates for a user
  subscribeToMatches(userId, callback) {
    const channels = [
      `databases.${DATABASE_ID}.collections.matching_suggestions.documents`,
      `user:${userId}:matches`,
      `user:${userId}:notifications`
    ];
    
    const matchCallback = (eventData) => {
      const { event, payload } = eventData;
      
      // Handle different types of match events
      if (event && payload) {
        let eventType = 'match_updated';
        
        // New match suggestions
        if (event.includes('matching_suggestions') && event.includes('.create')) {
          // Check if this user is involved in the match
          if (payload.user1_id === userId || payload.user2_id === userId) {
            eventType = 'new_match';
          }
        }
        // Match status updates
        else if (event.includes('matching_suggestions') && event.includes('.update')) {
          if (payload.user1_id === userId || payload.user2_id === userId) {
            if (payload.status === 'accepted') {
              eventType = 'match_accepted';
            } else if (payload.status === 'declined') {
              eventType = 'match_declined';
            }
          }
        }
        // Custom notification events
        else if (event.includes('new_match')) {
          eventType = 'new_match';
        }
        else if (event.includes('match_accepted')) {
          eventType = 'match_accepted';
        }
        
        // Only call callback for relevant events
        if (eventType) {
          callback({
            type: eventType,
            match: payload,
            ...eventData
          });
        }
      }
    };
    
    return this.subscribe(channels, matchCallback, `matches_${userId}`);
  }

  // Enhanced user notification subscription for AI matching
  subscribeToUserNotifications(userId, handlers = {}) {
    try {
      console.log('🔔 Setting up AI matching notifications for user:', userId);
      
      // Subscribe to match events
      const matchSubscriptionId = this.subscribeToMatches(userId, (eventData) => {
        const { type } = eventData;
        
        if (handlers[type]) {
          try {
            handlers[type](eventData);
          } catch (error) {
            console.error(`Error in ${type} handler:`, error);
          }
        }
        
        // Generic handler for all match events
        if (handlers.onMatch) {
          try {
            handlers.onMatch(eventData);
          } catch (error) {
            console.error('Error in onMatch handler:', error);
          }
        }
      });
      
      console.log('✅ AI matching notifications setup complete');
      return matchSubscriptionId;
      
    } catch (error) {
      console.error('❌ Failed to setup AI matching notifications:', error);
      return null;
    }
  }

  // Setup comprehensive real-time for a user
  setupUserRealtime(userId, callbacks = {}) {
    try {
      console.log('🔔 Setting up comprehensive real-time for user:', userId);
      
      const subscriptions = {};
      
      // Subscribe to messages
      if (callbacks.onMessage) {
        subscriptions.messages = this.subscribeToMessages(userId, callbacks.onMessage);
      }
      
      // Subscribe to trades
      if (callbacks.onTrade) {
        subscriptions.trades = this.subscribeToTrades(userId, callbacks.onTrade);
      }
      
      // Subscribe to AI matches
      if (callbacks.onMatch || callbacks.new_match || callbacks.match_accepted) {
        subscriptions.matches = this.subscribeToMatches(userId, callbacks.onMatch);
      }
      
      // Subscribe to marketplace updates
      if (callbacks.onMarketplace) {
        subscriptions.marketplace = this.subscribeToMarketplace(callbacks.onMarketplace);
      }
      
      // Subscribe to profile updates
      if (callbacks.onProfile) {
        subscriptions.profile = this.subscribeToUserUpdates(userId, callbacks.onProfile);
      }
      
      // Store user subscription group
      this.eventHandlers.set(`user_${userId}`, subscriptions);
      
      console.log('✅ User real-time setup complete:', Object.keys(subscriptions));
      
      return subscriptions;
      
    } catch (error) {
      console.error('❌ Setup user real-time failed:', error);
      return {};
    }
  }

  // Cleanup user real-time subscriptions
  cleanupUserRealtime(userId) {
    try {
      const subscriptions = this.eventHandlers.get(`user_${userId}`);
      
      if (subscriptions) {
        let count = 0;
        
        Object.values(subscriptions).forEach(subId => {
          if (this.unsubscribe(subId)) {
            count++;
          }
        });
        
        this.eventHandlers.delete(`user_${userId}`);
        
        console.log(`✅ Cleaned up ${count} user real-time subscriptions for user:`, userId);
        return count;
      }
      
      return 0;
      
    } catch (error) {
      console.error('❌ Cleanup user real-time failed:', error);
      return 0;
    }
  }

  // Send typing indicator (simulated real-time)
  sendTypingIndicator(senderId, recipientId, isTyping) {
    try {
      // In a real implementation, this would trigger a real-time event
      // For now, we'll simulate it with a custom event
      const event = new CustomEvent('typing-indicator', {
        detail: {
          senderId,
          recipientId,
          isTyping,
          timestamp: new Date().toISOString()
        }
      });
      
      document.dispatchEvent(event);
      
      console.log('💬 Typing indicator sent:', { senderId, recipientId, isTyping });
      return true;
      
    } catch (error) {
      console.error('❌ Send typing indicator failed:', error);
      return false;
    }
  }

  // Listen for typing indicators
  onTypingIndicator(callback) {
    const handler = (event) => callback(event.detail);
    document.addEventListener('typing-indicator', handler);
    
    // Return cleanup function
    return () => document.removeEventListener('typing-indicator', handler);
  }

  // Send read receipt (simulated real-time)
  sendReadReceipt(senderId, recipientId, messageId) {
    try {
      const event = new CustomEvent('read-receipt', {
        detail: {
          senderId,
          recipientId,
          messageId,
          timestamp: new Date().toISOString()
        }
      });
      
      document.dispatchEvent(event);
      
      console.log('👁️ Read receipt sent:', { senderId, recipientId, messageId });
      return true;
      
    } catch (error) {
      console.error('❌ Send read receipt failed:', error);
      return false;
    }
  }

  // Listen for read receipts
  onReadReceipt(callback) {
    const handler = (event) => callback(event.detail);
    document.addEventListener('read-receipt', handler);
    
    // Return cleanup function
    return () => document.removeEventListener('read-receipt', handler);
  }

  // Connection status
  checkConnectionStatus() {
    // In a real implementation, this would check the WebSocket connection
    return this.isConnected;
  }

  // Get active subscriptions info
  getActiveSubscriptions() {
    const subscriptions = [];
    
    for (const [id, subscription] of this.subscriptions) {
      subscriptions.push({
        id,
        channels: subscription.channels,
        createdAt: subscription.createdAt
      });
    }
    
    return subscriptions;
  }

  // Health check
  healthCheck() {
    return {
      isConnected: this.isConnected,
      activeSubscriptions: this.subscriptions.size,
      eventHandlers: this.eventHandlers.size,
      status: this.subscriptions.size > 0 ? 'active' : 'idle'
    };
  }

  // Event debugging
  enableDebugging() {
    this.debugMode = true;
    console.log('🐛 Real-time debugging enabled');
  }

  disableDebugging() {
    this.debugMode = false;
    console.log('🐛 Real-time debugging disabled');
  }
}

// Create and export singleton instance
const appwriteRealtime = new AppwriteRealtimeService();

// Auto-cleanup on page unload
window.addEventListener('beforeunload', () => {
  appwriteRealtime.unsubscribeAll();
});

// Named exports for convenience
export const subscribeToUserNotifications = (userId, handlers) => 
  appwriteRealtime.subscribeToUserNotifications(userId, handlers);

export const subscribeToMessages = (userId, callback) => 
  appwriteRealtime.subscribeToMessages(userId, callback);

export const subscribeToTrades = (userId, callback) => 
  appwriteRealtime.subscribeToTrades(userId, callback);

export const subscribeToMatches = (userId, callback) => 
  appwriteRealtime.subscribeToMatches(userId, callback);

export const subscribeToMarketplace = (callback, filters) => 
  appwriteRealtime.subscribeToMarketplace(callback, filters);

export default appwriteRealtime;

File: appwrite_realtime.py
"""
Appwrite Realtime Service for Trading Post
Handles real-time subscriptions, live updates, and notifications
"""

from appwrite.exception import AppwriteException
from appwrite_config import appwrite_config, DATABASE_ID, COLLECTIONS
import logging
from typing import Optional, Dict, Any, List, Callable
import asyncio
import json
from datetime import datetime

logger = logging.getLogger(__name__)


class AppwriteRealtimeService:
    """Handle real-time operations with Appwrite"""

    def __init__(self):
        self.client = appwrite_config.client
        self.active_subscriptions = {}
        self.event_handlers = {}

    def subscribe_to_collection(
        self, collection_id: str, callback: Callable, user_id: str = None, filters: List[str] = None
    ) -> str:
        """Subscribe to collection changes"""
        try:
            # Build subscription channels
            channels = [f"databases.{DATABASE_ID}.collections.{collection_id}.documents"]

            if user_id:
                # Subscribe to user-specific documents
                channels.append(f"databases.{DATABASE_ID}.collections.{collection_id}.documents.{user_id}")

            # Create subscription
            subscription_id = f"{collection_id}_{user_id or 'all'}_{len(self.active_subscriptions)}"

            def realtime_callback(response):
                try:
                    event_data = {
                        "event": response.get("event"),
                        "data": response.get("payload"),
                        "timestamp": datetime.utcnow().isoformat(),
                        "collection_id": collection_id,
                        "subscription_id": subscription_id,
                    }

                    # Apply filters if specified
                    if filters and not self._passes_filters(event_data, filters):
                        return

                    # Call the user-provided callback
                    callback(event_data)

                    logger.info(f"Realtime event processed: {response.get('event')}")

                except Exception as e:
                    logger.error(f"Realtime callback error: {e}")

            # Subscribe using Appwrite client
            unsubscribe = self.client.subscribe(channels, realtime_callback)

            # Store subscription
            self.active_subscriptions[subscription_id] = {
                "channels": channels,
                "callback": callback,
                "unsubscribe": unsubscribe,
                "created_at": datetime.utcnow().isoformat(),
                "collection_id": collection_id,
                "user_id": user_id,
            }

            logger.info(f"Subscribed to collection: {collection_id} - ID: {subscription_id}")
            return subscription_id

        except Exception as e:
            logger.error(f"Failed to subscribe to collection {collection_id}: {e}")
            return None

    def subscribe_to_user_messages(self, user_id: str, callback: Callable) -> str:
        """Subscribe to new messages for a user"""

        def message_callback(event_data):
            # Filter for messages where user is recipient
            message = event_data.get("data", {})
            if message.get("recipient_id") == user_id:
                callback(event_data)

        return self.subscribe_to_collection(COLLECTIONS["messages"], message_callback, filters=["create"])

    def subscribe_to_user_trades(self, user_id: str, callback: Callable) -> str:
        """Subscribe to trade updates for a user"""

        def trade_callback(event_data):
            # Filter for trades involving this user
            trade = event_data.get("data", {})
            if trade.get("user1_id") == user_id or trade.get("user2_id") == user_id:
                callback(event_data)

        return self.subscribe_to_collection(COLLECTIONS["trades"], trade_callback)

    def subscribe_to_marketplace_updates(self, callback: Callable, location: str = None, category: str = None) -> str:
        """Subscribe to marketplace listing updates"""

        def marketplace_callback(event_data):
            listing = event_data.get("data", {})

            # Apply location filter
            if location and listing.get("location") != location:
                return

            # Apply category filter
            if category and listing.get("category") != category:
                return

            callback(event_data)

        return self.subscribe_to_collection(COLLECTIONS["items"], marketplace_callback, filters=["create", "update"])

    def subscribe_to_user_matches(self, user_id: str, callback: Callable) -> str:
        """Subscribe to new AI matches for a user"""

        def match_callback(event_data):
            if event_data.get("event") == "databases.*.collections.*.documents.*.create":
                trade = event_data.get("data", {})
                if (trade.get("user1_id") == user_id or trade.get("user2_id") == user_id) and trade.get(
                    "status"
                ) == "ai_suggested":
                    callback(event_data)

        return self.subscribe_to_collection(COLLECTIONS["trades"], match_callback, filters=["create"])

    def unsubscribe(self, subscription_id: str) -> bool:
        """Unsubscribe from a subscription"""
        try:
            if subscription_id in self.active_subscriptions:
                subscription = self.active_subscriptions[subscription_id]

                # Call the unsubscribe function
                if subscription.get("unsubscribe"):
                    subscription["unsubscribe"]()

                # Remove from active subscriptions
                del self.active_subscriptions[subscription_id]

                logger.info(f"Unsubscribed from: {subscription_id}")
                return True

            return False

        except Exception as e:
            logger.error(f"Failed to unsubscribe {subscription_id}: {e}")
            return False

    def unsubscribe_all(self) -> int:
        """Unsubscribe from all active subscriptions"""
        count = 0
        subscription_ids = list(self.active_subscriptions.keys())

        for subscription_id in subscription_ids:
            if self.unsubscribe(subscription_id):
                count += 1

        logger.info(f"Unsubscribed from {count} subscriptions")
        return count

    def get_active_subscriptions(self) -> Dict[str, Any]:
        """Get list of active subscriptions"""
        return {
            sub_id: {
                "collection_id": sub["collection_id"],
                "user_id": sub["user_id"],
                "created_at": sub["created_at"],
                "channels": sub["channels"],
            }
            for sub_id, sub in self.active_subscriptions.items()
        }

    def broadcast_user_notification(self, user_id: str, notification_type: str, data: Dict[str, Any]) -> bool:
        """Send real-time notification to a specific user"""
        try:
            # In a real implementation, this would use Appwrite's messaging system
            # For now, we'll trigger callbacks for matching subscriptions

            notification = {
                "type": notification_type,
                "user_id": user_id,
                "data": data,
                "timestamp": datetime.utcnow().isoformat(),
            }

            # Find subscriptions for this user
            for sub_id, subscription in self.active_subscriptions.items():
                if subscription.get("user_id") == user_id:
                    try:
                        subscription["callback"](
                            {"event": "notification", "data": notification, "subscription_id": sub_id}
                        )
                    except Exception as e:
                        logger.error(f"Failed to deliver notification: {e}")

            logger.info(f"Notification broadcasted to user {user_id}: {notification_type}")
            return True

        except Exception as e:
            logger.error(f"Failed to broadcast notification: {e}")
            return False

    def broadcast_global_notification(
        self, notification_type: str, data: Dict[str, Any], target_groups: List[str] = None
    ) -> int:
        """Send real-time notification to multiple users"""
        count = 0

        try:
            notification = {
                "type": notification_type,
                "data": data,
                "timestamp": datetime.utcnow().isoformat(),
                "target_groups": target_groups,
            }

            # Broadcast to all active subscriptions or filtered groups
            for sub_id, subscription in self.active_subscriptions.items():
                try:
                    # Apply group filtering if specified
                    if target_groups:
                        user_groups = subscription.get("groups", [])
                        if not any(group in target_groups for group in user_groups):
                            continue

                    subscription["callback"](
                        {"event": "global_notification", "data": notification, "subscription_id": sub_id}
                    )
                    count += 1

                except Exception as e:
                    logger.error(f"Failed to deliver global notification: {e}")

            logger.info(f"Global notification broadcasted to {count} subscribers: {notification_type}")
            return count

        except Exception as e:
            logger.error(f"Failed to broadcast global notification: {e}")
            return 0

    def _passes_filters(self, event_data: Dict[str, Any], filters: List[str]) -> bool:
        """Check if event data passes the specified filters"""
        try:
            event_type = event_data.get("event", "")

            for filter_condition in filters:
                if filter_condition == "create" and "create" in event_type:
                    return True
                elif filter_condition == "update" and "update" in event_type:
                    return True
                elif filter_condition == "delete" and "delete" in event_type:
                    return True
                elif filter_condition in event_type:
                    return True

            return len(filters) == 0  # No filters means pass all

        except Exception:
            return False

    # Convenience methods for common real-time scenarios
    def setup_user_realtime(self, user_id: str, callbacks: Dict[str, Callable]) -> Dict[str, str]:
        """Setup complete real-time subscriptions for a user"""
        subscriptions = {}

        try:
            # Subscribe to messages
            if "messages" in callbacks:
                sub_id = self.subscribe_to_user_messages(user_id, callbacks["messages"])
                if sub_id:
                    subscriptions["messages"] = sub_id

            # Subscribe to trades
            if "trades" in callbacks:
                sub_id = self.subscribe_to_user_trades(user_id, callbacks["trades"])
                if sub_id:
                    subscriptions["trades"] = sub_id

            # Subscribe to matches
            if "matches" in callbacks:
                sub_id = self.subscribe_to_user_matches(user_id, callbacks["matches"])
                if sub_id:
                    subscriptions["matches"] = sub_id

            # Subscribe to profile updates
            if "profile" in callbacks:
                sub_id = self.subscribe_to_collection(COLLECTIONS["users"], callbacks["profile"], user_id)
                if sub_id:
                    subscriptions["profile"] = sub_id

            logger.info(f"Setup {len(subscriptions)} real-time subscriptions for user {user_id}")
            return subscriptions

        except Exception as e:
            logger.error(f"Failed to setup user realtime: {e}")
            return subscriptions

    def cleanup_user_realtime(self, subscription_ids: Dict[str, str]) -> int:
        """Cleanup user's real-time subscriptions"""
        count = 0

        for sub_type, sub_id in subscription_ids.items():
            if self.unsubscribe(sub_id):
                count += 1

        return count

    async def send_typing_indicator(self, sender_id: str, recipient_id: str, is_typing: bool) -> bool:
        """Send typing indicator in real-time"""
        try:
            return self.broadcast_user_notification(
                recipient_id, "typing_indicator", {"sender_id": sender_id, "is_typing": is_typing}
            )

        except Exception as e:
            logger.error(f"Failed to send typing indicator: {e}")
            return False

    async def send_read_receipt(self, sender_id: str, recipient_id: str, message_id: str) -> bool:
        """Send read receipt in real-time"""
        try:
            return self.broadcast_user_notification(
                sender_id, "read_receipt", {"recipient_id": recipient_id, "message_id": message_id}
            )

        except Exception as e:
            logger.error(f"Failed to send read receipt: {e}")
            return False


# Global instance
appwrite_realtime = AppwriteRealtimeService()
